### PR TITLE
Remove the references to `_file_obj` outside low level code paths, change to `_close`

### PR DESCRIPTION
**Pull request by alexamici**
- [x] Related to #4309
- [x] No tests added, all tests passing
- [x] Passes `pre-commit run --all-files`
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [x] No new functions/methods are listed in `api.rst`

You can now pass a `close` argument to `xr.Dataset` that closes the backend resources. This is only useful to backend developers, we need to add the documentation as per #4803 but it may need to be hidden and not documented to end-users. 

The PR is draft pending some cleanups and the documentation.

### CODE REVIEW

**Comment by shoyer** [comment]
Per discussion, let's create a `set_close` method instead of adding this into the Dataset/DataArray constructor.

File: xarray/core/common.py [hunk 1]
```diff
@@ -1264,9 +1266,11 @@ def where(self, cond, other=dtypes.NA, drop: bool = False):
 
         return ops.where_method(self, cond, other)
 
+    def set_close(self, close: Optional[Callable[[], None]]) -> None:
```
Comment (shoyer) [inline xarray/core/common.py:1269]: Can you add a short docstring?

File: xarray/core/dataarray.py [hunk 2]
```diff
@@ -421,22 +422,29 @@ def __init__(
...
         if name is _default:
             name = self.name
-        return type(self)(variable, coords, name=name, fastpath=True, indexes=indexes)
+        if close is _default:
+            close = self._close
+        replaced = type(self)(
+            variable, coords, name=name, fastpath=True, indexes=indexes
+        )
+        replaced.set_close(close)
+        return replaced
```
Comment (shoyer) [inline xarray/core/dataarray.py:447]: I would prefer to revert this change to `_replace`, and require users to always explictly call `set_close` rather than to update it automatically in only some cases.

Comment (alexamici) [reply xarray/core/dataarray.py:447]: Not sure I understand: `_close` follows the same rules as any other `Dataset` / `DataArray` attribute, it is updated if passed otherwise it is copied from `self`.

Note that this change is the main reason the PR is so small. If I remove this I'll need to hunt down all the calls to `_replace` and choose whether to add or not to add `set_close`.

I may not have understood your concern.

Comment (shoyer) [reply xarray/core/dataarray.py:447]: > Not sure I understand: _close follows the same rules as any other Dataset / DataArray attribute, it is updated if passed, otherwise it is copied from self.

I'm suggesting that `_close` should be an exception to this rule, and that we should never call `set_close()` inside `_replace()`. Anytime `_replace()` is called (creating a new `DataArray`), the right thing to do is to lose the `close` method.

> If I remove this I'll need to hunt down all the calls to _replace and choose whether to add or not to add set_close.

My suggestion is that it is OK to *never* add `set_close` (unless inside a function like `open_dataset` or `open_mfdataset` that opens file object(s) and already assigns to `_file_obj`)

Comment (alexamici) [reply xarray/core/dataarray.py:447]: So the `Dataset.close` method and `DataArray.close` method are supposed to actually release the resources only when called from the objects returned directly from the `xr.open_*` functions?

```python
>>> ds1 = xr.open_dataset("file.nc")
>>> ds1.close()  # releases the resources

>>> ds2 = xr.open_dataset("file.nc").rename_dims(valid_time="time")
>>> ds2.close()  # doesn't release the resources
```

I don't understand the rationale, but I'll trust your judgment on this.

Comment (alexamici) [reply xarray/core/dataarray.py]: Done, this is ready for me.

**Review by shoyer — APPROVED** [review]

File: xarray/core/dataarray.py (hunk 2 above)
Comment (kmuehlbauer) [reply xarray/core/dataarray.py]: @shoyer This is really great and helps a lot in resource handling. Your example should be put into documentation.

File: xarray/core/common.py [hunk 3]
```diff
@@ -1263,11 +1266,24 @@ def where(self, cond, other=dtypes.NA, drop: bool = False):
 
         return ops.where_method(self, cond, other)
 
+    def set_close(self, close_store: Optional[Callable[[], None]]) -> None:
```
Comment (shoyer) [inline xarray/core/common.py:1269]: Maybe just call this argument `close` rather than `close_store`? In principle, this is very generic and not "store" specific.

File: xarray/backends/api.py [hunk 4]
```diff
@@ -963,7 +952,11 @@ def open_mfdataset(
             ds.close()
         raise
 
-    combined._file_obj = _MultiFileCloser(file_objs)
+    def multi_file_closer():
+        for closer in closers:
+            closer()
+
+    combined._close = multi_file_closer
```
[the remark this answers is not in the document]
Comment (alexamici) [reply xarray/backends/api.py:959]: d'oh!

File: xarray/core/common.py [hunk 5]
```diff
@@ -1263,11 +1266,24 @@ def where(self, cond, other=dtypes.NA, drop: bool = False):
...
+        """Register the function that releases all resources used by the data store.
+
+        This method is mostly intended for backend developers and it is rarely
+        needed by regular end-users.
```
Comment (shoyer) [inline xarray/core/common.py:1273]:
```suggestion
        This method controls how xarray cleans up resources associated
        with this object when the .close() method is called. It is mostly
        intended for backend developers and it is rarely needed by regular
        end-users.
```

**Comment by alexamici** [comment]
I'll merge it after the last cleanup as I need this to start working on the API v1 to v2 PR. Thanks @shoyer !

File: xarray/core/common.py [hunk 6]
```diff
@@ -1263,11 +1266,24 @@ def where(self, cond, other=dtypes.NA, drop: bool = False):
...
+
+        Parameters
+        ----------
+        close_store : callable
+            A callable that releases all resources used by the data store.
```
Comment (shoyer) [inline xarray/core/common.py:1278]:
```suggestion
            A callable that when called like ``close_store()`` releases
            all resources used by the data store.
```

**Review by shoyer — APPROVED** [review]

**Comment by alexamici** [comment]
woops, I see you didn't finish, yet :)

File: xarray/core/common.py (hunk 3 above)
Comment (alexamici) [reply xarray/core/common.py]: Agreed. Done.

File: xarray/core/common.py [hunk 7]
```diff
@@ -1263,11 +1266,27 @@ def where(self, cond, other=dtypes.NA, drop: bool = False):
 
         return ops.where_method(self, cond, other)
 
+    def set_close(self, close: Optional[Callable[[], None]]) -> None:
+        """Register the function that releases all resources used by the data store.
```
Comment (alexamici) [inline xarray/core/common.py:1270]: @shoyer not `set_close` and `close` use slightly different wording:

close -> `Close any files linked to this object`

set_close -> `Register the function that releases all resources used by the data store.`

I would go with `release(s) any resources linked to this object` everywhere, but I'm not a native speaker and I'm generally not very good a writing user documentation.

The important point is that we use the same working everywhere.

Comment (shoyer) [reply xarray/core/common.py:1270]: I like `release(s) any resources linked to this object`
